text/0000-assoc-default-groups.md

```diff
@@ -0,0 +1,1733 @@
+today would be accepted.
+
+Meanwhile, if `impl Iterator` meant `impl Iterator<Item = ()>`,
+this would impose a stronger requirement on existing code where `impl Iterator`
+is used and thus it would be a breaking change to the users of `Iterator`.
+
+For `Iterator`, it would not be helpful to introduce a default for `Item`.
+However, for the purposes of API evolution, the value is not in assigning
+defaults to the existing associated types of a trait. Rather, the value comes
+from being able to add associated types without breaking dependent crates.
+
+Due to the possible breakage of `dyn Trait<..>` when adding an associated type
+to `Trait`, to truly achieve API evolution, defaults must be taken into account
+and be inferable for `dyn Trait`. The opposite is true for `impl Trait`.
+To facilitate API evolution, stronger requirements must not be placed on
+`impl Trait` and therefore defaults should not be taken into account.
+
+# Prior art
+[prior-art]: #prior-art
+
+## Haskell
+
+[associated type defaults]: https://www.microsoft.com/en-us/research/wp-content/uploads/2005/01/at-syns.pdf
+
+As Rust traits are a form of type classes,
+we naturally look for prior art from were they first were introduced.
+That language, being Haskell,
+permits a user to specify [associated type defaults].
+For example, we may write the following legal program:
+
+```haskell
+{-# LANGUAGE TypeFamilies #-}
+
+class Foo x where
+  type Bar x :: *
+  -- A default:
+  type Bar x = Int
+
+  -- Provided method:
+  baz :: x -> Bar x -> Int
+  baz _ _ = 0
+
+data Quux = Quux
+
+instance Foo Quux where
+  baz _ y = y
+```
+
+As in this proposal, we may assume that `y :: Int` in the above snippet.
+
+In this case, we are not assuming that `Bar x` unifies with `Int` in the `class`.
+Let's try to assume that now:
+
+```haskell
+{-# LANGUAGE TypeFamilies #-}
+
+class Foo x where
+  type Bar x :: *
+  -- A default:
+  type Bar x = Int
+
+  -- Provided method:
+  baz :: x -> Bar x -> Int
+  baz _ barX = barX
+```
+
+This snippet results in a type checking error (tested on GHC 8.0.1):
+
+```
+main.hs:11:16: error:
+    • Couldn't match expected type ‘Int’ with actual type ‘Bar x’
+    • In the expression: barX
+      In an equation for ‘baz’: baz _ barX = barX
+    • Relevant bindings include
+        barX :: Bar x (bound at main.hs:11:9)
+        baz :: x -> Bar x -> Int (bound at main.hs:11:3)
+<interactive>:3:1: error:
+```
+
+The thing to pay attention to here is:
+> Couldn't match expected type ‘`Int`’ with actual type ‘`Bar x`’
+
+We can clearly see that the type checker is *not* allowing us to assume
+that `Int` and `Bar x` are the same type.
+This is consistent with the approach this RFC proposes.
+
+To our knowledge, Haskell does not have any means such as `default { .. }`
+to change this behaviour. Presumably, this is the case because Haskell
+preserves parametricity thus lacking specialization,
+wherefore `default { .. }`, as suggested in the [future possibilities],
```

Comment:
nit: the `[future possibilities]` link is broken

Reply:
Thanks, fixed!